### ISSUE TITLE
Support for different namespaces in the same kubernetes cluster.

### ISSUE THREAD

**Issue by khteh** [kubernetes-sigs/aws-alb-ingress-controller#849]
Hi,
   I have had a production cluster running on AWS EKS with ALB Ingress Controller. I am planning my application to support multi-tenant and therefore I think of namespace. I plan to use different namespaces for different customer sharing the same underlying kubernetes infrastructure (nodes, auto-scaling, etc) for URLs like customer-application.mydomain.com. I would like to know how does it work with ALB ingress controller in terms of scalability / maitainability? Do I have to use separate ingress controllers or a single ingress controller with multiple target groups? Thanks.

**Comment by M00nF1sh** [comment]
Hi, 
Generally an single ingress controller for the whole cluster is sufficient.(It's event driving by endpoints/service/ingress updates).

Also, we do support to run separate ingress controller per namespace, below are the factors i can think of.
1. version control:
    1. Pros: you can have separate version control on ingress controller itself for different customers. e.g. avoid an version upgrade of ingress controller for customerA break customerB. 
    1. Cons: maintain the different versions of ingress controller itself is an maintainability burden. We'll do our best to maintain backwards compability, and will have an *action required* section in release notes if there are any(we may release scripts to auto-detect and upgrades). So maintain a single ingress controller for whole cluster is easier.

1. responsiveness:
    Currently the ingress controller runs a work queue of a single worker thread. So ingresses are updated sequentially. 
    1. Pros: running separate ingress controller will allow changes be picked up and resonciled quicker.
    1. Cons: However, we are working on several features like enable multiple worker thread, and limiting reconcile scope(e.g. endpoints changes only affect targetGroups without check listener settings), so that the responsiveness for a single ingress controller for whole cluster will be equally good enough.

In conclusion, I recommend to running a single ingress controller per cluster to improve maintainability, and working to close the gap.

**Comment by khteh** [comment]
Hi, how do I specify different namespaces in the same ingress yaml file for the ingress controller? Thanks.

**Comment by khteh** [comment]
Or in different file sharing the same ingress controller...

**Comment by kuengnee** [comment]
Hi,

just to addon to the question. how about maintainability and scalability? When new client aboard, an new namespace is created. Then the ingress yml need to update with the new config with a new host. 

Wont it be hard to scale under this approach?

**Comment by M00nF1sh** [comment]
@khteh @kuengnee 
Do you mean share the same ALB? ALB ingress controller behavies differently from nginx ingress controller. In nginx ingress controller, one controller corresponding to a nginx instance and service(and all ingress rules are aggregated to the nginx instance).
In ALB ingress controller, the number of ALB isn't tied to the ingress controller itself(so we don't need multiple ALB ingress controller). Currently we provision one ALB per ingress objects, and i'm working on a solution to allow you aggregate ingress objects. Basically grants you flexibility on how to provision ALBs.  
As for scalability, as long as the of ingress/service objects in your k8s cluster can fit into controller memory(we keep a local copy of these objects to avoid overwhelm APIServer), we are good to run a single ALB ingress controller for whole cluster.

**Comment by khteh** [comment]
Hi @M00nF1sh, we were worried about maintaining the ever-growing of the ingress manifest yaml file....

**Comment by M00nF1sh** [comment]
Hi, we are going to support you maintain your ingress file in a distributed way soon(planed to release in March)
